### PR TITLE
"Get" a structured header from a header list.

### PR DESCRIPTION
**Pull request by mikewest**
As discussed in #930 and mikewest/corpp#2, this patch adds a "get"
method on header lists that allows consistent extraction of
structured headers.

***
<a href="https://whatpr.org/fetch/983.html" title="Last updated on Dec 21, 2019, 8:54 AM UTC (9125901)">Preview</a> | <a href="https://whatpr.org/fetch/983/b93383c...9125901.html" title="Last updated on Dec 21, 2019, 8:54 AM UTC (9125901)">Diff</a>

### CODE REVIEW

File: fetch.bs [hunk 1]
```diff
@@ -584,10 +585,32 @@ pair in a <a for=/>header list</a> <var>list</var>, run these steps:
...
-objects in <a for=/>header lists</a> by serializing them. In the future the fact that they are
-objects might be preserved end-to-end. [[!HEADER-STRUCTURE]]
+<p>To
+<dfn export for="header list" id=concept-header-list-get-structured-header>get a structured header</dfn>
+given a <var>name</var> and a <var>type</var> from a <a for=/>header list</a> <var>list</var>, run
+these steps:
+
+<ol>
+ <li><p>If <var>type</var> is not one of "<code>dictionary</code>", "<code>list</code>", or
+ "<code>item</code>", return failure.
```
Comment (annevk) [inline fetch.bs]: This should probably be an assert, right? At least until we have an API and at that point we could use an enum on the IDL boundary to enforce it.

File: fetch.bs [hunk 2]
```diff
@@ -584,10 +585,32 @@ pair in a <a for=/>header list</a> <var>list</var>, run these steps:
...
+ <li><p>Let <var>value</var> be the result of <a for="header list">getting</a> <var>name</var> from
+ <var>list</var>.
+ <li><p>If <var>value</var> is null, return null.
```
Comment (annevk) [inline fetch.bs]:
```suggestion
 <li><p>If <var>value</var> is null, then return null.
```

File: fetch.bs [hunk 3]
```diff
@@ -584,10 +585,32 @@ pair in a <a for=/>header list</a> <var>list</var>, run these steps:
...
+ algorithm with <var ignore>input_string</var> set to <var>value</var>, and
+ <var ignore>header_type</var> set to <var>type</var>.
+ <li><p>If parsing failed, return failure.
+ <li><p>Return <var>result</var>.
+</ol>
+
+<p class=note><a>Structured header values</a> are defined as objects which HTTP can (eventually)
+serialize in interesting and efficient ways. For the moment, Fetch only supports <a for=/>header</a>
+<a for=header>values</a> as <a for=/>byte sequences</a>, which means that these objects can be set
+in <a for=/>header lists</a> only via serialization, and they can be obtained from <a for=/>header
```
Comment (annevk) [inline fetch.bs]: Please keep elements like `<a>` on a single line.

File: fetch.bs [hunk 4]
```diff
@@ -584,10 +585,32 @@ pair in a <a for=/>header list</a> <var>list</var>, run these steps:
...
+<ol>
+ <li><p>If <var>type</var> is not one of "<code>dictionary</code>", "<code>list</code>", or
+ "<code>item</code>", return failure.
+ <li><p>Let <var>value</var> be the result of <a for="header list">getting</a> <var>name</var> from
+ <var>list</var>.
+ <li><p>If <var>value</var> is null, return null.
+ <li><p>Let <var>result</var> be the result of executing the <a>parsing structured headers</a>
+ algorithm with <var ignore>input_string</var> set to <var>value</var>, and
+ <var ignore>header_type</var> set to <var>type</var>.
+ <li><p>If parsing failed, return failure.
```
Comment (annevk) [inline fetch.bs]:
```suggestion
 <li><p>If parsing failed, then return failure.
```

File: fetch.bs [hunk 5]
```diff
@@ -584,10 +585,32 @@ pair in a <a for=/>header list</a> <var>list</var>, run these steps:
...
+</ol>
+
+<p class=note><a>Structured header values</a> are defined as objects which HTTP can (eventually)
+serialize in interesting and efficient ways. For the moment, Fetch only supports <a for=/>header</a>
+<a for=header>values</a> as <a for=/>byte sequences</a>, which means that these objects can be set
+in <a for=/>header lists</a> only via serialization, and they can be obtained from <a for=/>header
+lists</a> only by parsing. In the future the fact that they are objects might be preserved
+end-to-end. [[!HEADER-STRUCTURE]]
+
+</div>
```
Comment (annevk) [inline fetch.bs]: It seems this and the newline before it were added accidentally? Or is this fixing an existing mistake?

**Comment by zcorpan** [comment]
> Thanks, so the one question I have is whether we want to further map a structured header value to an Infra value?

I think that seems nice. But how would we differentiate between token and string?

**Comment by annevk** [comment]
We'd have a struct consisting of type, value, and parameters, I suspect. (We'd need something similar for the JavaScript API.) But I'm okay leaving that for later and refactoring things as appropriate.

(An alternative that might be nice is that we define implicit mapping, similar to how we deal with various IDL concepts.)

**Comment by domenic** [comment]
I'm really confused why we'd introduce this wrapper. All it seems to add over using the SH spec directly is an assert and a redundant "if failure, return failure".

In https://wicg.github.io/origin-policy/#parse-an-origin-policy-header I used the SH spec directly and I see no compelling reason to change.

I am -1 on this PR.

**Comment by annevk** [comment]
Well, you could reduce the caller algorithm slightly because of this as you no longer have to grab the header value separately. Fetch contains a lot of algorithms like that which make the ergonomics better and generally makes specifications easier to read.

Also, if at some point a header list is able to contain header values that do not need to be parsed because they are already of the correct type, we'd have one place to change to account for that rather than a dozen.

**Comment by mikewest** [comment]
Like @annevk, I see this patch as creating a single point we can change in the future if/when we have cleaner support for structured headers as objects in themselves that are persisted end-to-end. I didn't hang type information in this patch, but I think it's probably worth doing in a subsequent PR.

File: fetch.bs (hunk 3 above)
Comment (mikewest) [reply fetch.bs]: Done.

File: fetch.bs (hunk 5 above)
Comment (mikewest) [reply fetch.bs]: Accidental, thanks!

File: fetch.bs (hunk 1 above)
Comment (mikewest) [reply fetch.bs]: Changed to an assert.

**Comment by annevk** [comment]
Pushed a hopefully final nit to this branch. Happy to merge it now but will wait a bit to see if there's final comments.

**Comment by mikewest** [comment]
Still LGTM, FWIW. Happy holidays! :)

**Comment by annevk** [comment]
Happy holidays to you all too!

**Pull request merged**